File: leetcode.com/reverse-linked-list.py
#!/bin/python3
# https://leetcode.com/problems/reverse-linked-list/

from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = None
        next = head

        while next:
            next_node = next.next
            next.next = current

            current = next
            next = next_node

        return current

Solution().reverseList(ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))

File: leetcode.com/reverse-linked-list.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end
# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
    return head if head&.next.nil?

    previous = head
    current = head.next
    nxt = current&.next

    head.next = nil
    
    while !current.nil? do
        t_nxt = nxt.next if nxt
        nxt.next = current if nxt
        current.next = previous

        previous = current
        current = nxt
        nxt = t_nxt
    end

    return previous
end